backend/fms_core/services/sample.py

```diff
@@ -443,10 +443,88 @@ def remove_qc_flags(sample):
         sample.quality_flag = None
         sample.save()
     except Error as e:
-        errors.appends(';'.join(e.messages))
+        errors.append(';'.join(e.messages))
 
     return sample, errors, warnings
 
 
+def add_sample_metadata(sample, metadata):
+    errors = []
+    warnings = []
+
+    if sample and metadata:
+        try:
+            # Retrieve the biosample of the given sample
+            biosample_obj = sample.biosample_not_pool
+            for name in metadata.keys():
+                value = metadata[name]
+                # Check if sample has already the metadata
+                if SampleMetadata.objects.filter(name=name, biosample=biosample_obj).exists():
+                    errors.append(f'Sample [{sample.name}] already has property [{name}] with value [{value}].')
+                else:
+                    SampleMetadata.objects.create(name=name, value=value, biosample=biosample_obj)
+        except ValidationError as e:
+            errors.append(';'.join(e))
+    else:
+        errors.append('Sample and metadata are required')
+
+    return metadata, errors, warnings
+
+
+def update_sample_metadata(sample, metadata):
+    errors = []
+    warnings = []
+
+    if sample and metadata:
+        try:
+            # Retrieve the biosample of the given sample
+            biosample_obj = sample.biosample_not_pool
+            for name in metadata.keys():
+                value = metadata[name]
+                # Check if sample has already the metadata
+                if SampleMetadata.objects.filter(name=name, biosample=biosample_obj).exists():
+                    metadata_obj = SampleMetadata.objects.get(name=name, biosample=biosample_obj)
+                    # Add warning if the new value is the same as the old value
+                    if metadata_obj.value == value:
+                        warnings.append(f'Sample [{sample.name}] has metadata [{name}] with the same value [{value}]')
+                    metadata_obj.value = value
+                    metadata_obj.save()
+                else:
+                    errors.append(f'Sample [{sample.name}] does not have metadata with name [{name}].')
+        except ValidationError as e:
+            errors.append(';'.join(e))
+    else:
+        errors.append('Sample and metadata are required')
+
+    return metadata, errors, warnings
+
+
+def remove_sample_metadata(sample, metadata):
+    num_objects_deleted = 0
+    errors = []
+    warnings = []
+
+    if sample and metadata:
+        try:
+            # Retrieve the biosample of the given sample
+            biosample_obj = sample.biosample_not_pool
+            for name in metadata.keys():
+                value = metadata[name]
+                metadata_obj = SampleMetadata.objects.get(name=name, biosample=biosample_obj)
+                # Add warning if the new value is the same as the old value
+                if metadata_obj.value != value:
```

Comment:
I am not sure this will be sufficient and safe for this functionality. I am afraid someone will mix ADD, UPDATE and DELETE in a template. I would force them to type in "DELETE" in the deleted column to make sure or we could prevent them from mixing actions.